lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -262,6 +284,15 @@ DocIdSet get(Query key, IndexReader.CacheHelper cacheHelper) {
     assert lock.isHeldByCurrentThread();
     assert key instanceof BoostQuery == false;
     assert key instanceof ConstantScoreQuery == false;
+
+    /*
+     * If the current query is already being asynchronously cached,
+     * do not trigger another cache operation
+     */
+    if (inFlightAsyncLoadQueries.contains(key)) {
```

Comment:
Hmm can we fix the concurrency issue here, e.g. just call `inFlightAsyncLoadQueries.add` and if the returned value is `false` (because the query was already in the set) then throw the exception?  Or, move this check down into the `cacheAsynchronously` method where we are doing the `add`?

Reply:
Good idea, fixed, thanks